vision/datasets/imageloader.go

```diff
@@ -10,74 +10,100 @@ import (
 	"github.com/wangkuiyi/gotorch/vision/transforms"
 )
 
-// ImageLoader struct
-type ImageLoader struct {
-	r      *tgz.Reader
-	vocab  map[string]int64
-	err    error
-	inputs []torch.Tensor
-	labels []int64
-	trans  *transforms.ComposeTransformer
-	mbSize int
+type sample struct {
+	data  image.Image
+	label int
 }
 
-func (p *ImageLoader) tensorGC() {
-	p.inputs = []torch.Tensor{}
-	p.labels = []int64{}
-	torch.GC()
+// ImageLoader struct
+type ImageLoader struct {
+	r       *tgz.Reader
+	vocab   map[string]int
+	samples chan sample
+	err     chan error
+	trans1  *transforms.ComposeTransformer // transforms before `ToTensor`
+	trans2  *transforms.ComposeTransformer // transforms after and include `ToTensor`
+	mbSize  int
+	inputs  []torch.Tensor
+	labels  []int64
 }
 
 // NewImageLoader returns an ImageLoader
-func NewImageLoader(fn string, vocab map[string]int64, trans *transforms.ComposeTransformer, mbSize int) (*ImageLoader, error) {
+func NewImageLoader(fn string, vocab map[string]int, trans *transforms.ComposeTransformer, mbSize int) (*ImageLoader, error) {
 	r, e := tgz.OpenFile(fn)
 	if e != nil {
 		return nil, e
 	}
-	return &ImageLoader{
-		r:      r,
-		vocab:  vocab,
-		err:    nil,
-		trans:  trans,
-		mbSize: mbSize,
-	}, nil
+	trans1, trans2 := splitComposeByToTensor(trans)
+	m := &ImageLoader{
+		r:       r,
+		vocab:   vocab,
+		samples: make(chan sample, mbSize*4),
```

Comment:
Do we need a buffered channel?

Reply:
I think so, the buffered channel can get better performance,  `dataloade`r can reading data at training.